adder46/pyhstr
# Add IPython support

## Description

These changes allow the application to work within IPython as it does in the classic Python shell. It leverages IPython's [custom magic](https://ipython.readthedocs.io/en/stable/config/custommagics.html) instead of passing by the display hook -- this results in a much simpler integration as toying with `sys.displayhook` in IPython is a bit thorny.

Using line magic, the history application can be called via `%ihh` and results in the same interactivity as using `hh` in the regular shell. Since custom magics are registered on import, a check to see if `IPython` exists as an importable package gates it; this way, we avoid wonky behaviour when `IPython` is not available.

## QA

- Spun up a venv, packaged and installed `pyhstr` but not `IPython`.
- Tried out the regular command (i..e `python -ic "from pyhstr import hh"`) and verified that `hh` works as expected in the shell.
- Spun up another venv, this time with `pyhstr` and `IPython`.
- Started `IPython` via `ipython -i -c="import pyhstr"`.
- Called `%ihh` from within the `IPython` shell.
- Verified that behaviour is similar to what happens in the regular shell.
